python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,179 @@
+
+from iceberg.io.base import FileIO, InputFile, InputStream, OutputFile, OutputStream
+
+
+class PyArrowFile(InputFile, OutputFile):
+    """A combined InputFile and OutputFile implementation that uses a pyarrow filesystem to generate pyarrow.lib.NativeFile instances
+
+    Args:
+        location(str): A URI or a path to a local file
+
+    Attributes:
+        location(str): The URI or path to a local file for a PyArrowFile instance
+        exists(bool): Whether the file exists or not
+        filesystem(pyarrow.fs.FileSystem): An implementation of the FileSystem base class inferred from the location
+
+    Examples:
+        >>> from iceberg.io.pyarrow import PyArrowFile
+        >>> input_file = PyArrowFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()  # Read the contents of the PyArrowFile instance
+        >>> output_file = PyArrowFile("s3://baz/qux.txt")
+        >>> output_file.create().write(b'foobytes')  # Write bytes to the PyArrowFile instance
+    """
+
+    def __init__(self, location: str):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+        if not parsed_location.scheme:  # If no scheme, assume the path is to a local file
+            self._filesystem, self._path = FileSystem.from_uri(os.path.abspath(location))
+        else:
+            self._filesystem, self._path = FileSystem.from_uri(location)  # Infer the proper filesystem
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file_info = self._filesystem.get_file_info(self._path)
+        return file_info.size
+
+    def exists(self) -> bool:
+        """Checks whether the location exists
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error such as
+                an AWS error code 15
+        """
+        try:
+            file_info = self._filesystem.get_file_info(self._path)
+        except OSError as e:
+            if e.errno == 13 or "AWS Error [code 15]" in str(e):
+                raise PermissionError(f"Cannot check if file exists, access denied: {self.location}")
+            raise  # pragma: no cover - If some other kind of OSError, raise the raw error
+        return False if file_info.type == FileType.NotFound else True
+
+    def open(self) -> InputStream:
+        """Opens the location using a PyArrow FileSystem inferred from the location
+
+        Returns:
+            pyarrow.lib.NativeFile: A NativeFile instance for the file located at `self.location`
+
+        Raises:
+            FileNotFoundError: If the file at self.location does not exist
+            PermissionError: If the file at self.location cannot be accessed due to a permission error such as
+                an AWS error code 15
+        """
+        try:
+            input_file = self._filesystem.open_input_file(self._path)
+        except OSError as e:
```

Comment:
Rather than checking `isinstance(FileNotFoundError)` I think it would be better to throw the original. That way you don't have the longer error within an error.

This should work:

```python
try:
    ...
except FileNotFoundError:
    raise
except PermissionError:
    raise
except OSError as e:
    if e.errno == 2 or "Path does not exist" in str(e):
        raise FileNotFoundError(...)
    ...
```

Reply:
Updated!